_rules/element-lang-programmatically-determinable-7ed469.md

```diff
@@ -0,0 +1,180 @@
+rule_type: atomic
+description: |
+  This rule checks that the language of an element in the page body can be programmatically determined.
+accessibility_requirements: # Remove whatever is not applicable
+  wcag20:3.1.2: # Language of Parts (AA)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing is needed
+    inapplicable: further testing is needed
+input_aspects:
+  -  DOM Tree
+  -  CSS Styling
+
+acknowledgements:
+  authors:
+    -  Patrick H. Lauke
+---
+
+## Applicability
+
+This rules applies to any HTML element that:
+
+* contains a [text node](https://dom.spec.whatwg.org/#text) as a [child](https://dom.spec.whatwg.org/#concept-tree-child) element
+* is a [descendant](https://dom.spec.whatwg.org/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) of a `body` element
+* is [included in the accessibility tree][]
+
+## Expectation (1)
+
+For each test target, the test target itself or any of its [ancestor](https://dom.spec.whatwg.org/#concept-tree-ancestor) elements (up to and including the root element of the document) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) has a `lang` attribute with a valid language subtag.
+
+## Assumptions
+
+_There are currently no assumptions_
+
+## Accessibility Support
+
+_There are no major accessibility support issues known for this rule._
+
+## Background
+
+In practice, this rule will be satisfied if the separate success criterion [3.1.1 Language of Page (Level: A)](https://www.w3.org/TR/WCAG21/#language-of-page) is satisfied. This rule ensures that if 3.1.1 is *not* satisfied, content is checked for alternative ways in which language of page content itself can still have a programmatically determinable human language.
+
+- [CSS Scoping Module Level 1 (editor's draft)](https://drafts.csswg.org/css-scoping/)
+- [Understanding Success Criterion 3.1.2: Language of Parts](https://www.w3.org/WAI/WCAG21/Understanding/language-of-parts.html)
+
+## Test Cases
+
+### Passed
+
+#### Passed Example 1
+
+The language of the `p` element is inherited by the `lang` attribute on the `html` root element.
+
+```html
+<html lang="en">
+	<body>
+		<p>Content</p>
+	</body>
+</html>
+```
+
+#### Passed Example 2
+
+The language of the `p` element is inherited by the `lang` attribute on the `div` parent element. Note that this example fails 3.1.1 Language of Page (Level: A) as the `html` root element lacks a `lang` attribute.
+
+```html
+<html>
+	<body>
+		<div lang="en">
+			<p>Content</p>
+		</div>
+	</body>
+</html>
+```
+
+#### Passed Example 3
+
+The language of the `p` element is inherited by the `lang` attribute on the `div` ancestor element. Note that this example fails 3.1.1 Language of Page (Level: A) as the `html` root element lacks a `lang` attribute.
+
+```html
+<html>
+	<body>
+		<div lang="en">
+			<div>
+				<p>Content</p>
+			</div>
+		</div>
+	</body>
+</html>
+```
```

Comment:
These two examples are actually pretty much the same (the `lang` attribute is on an ancestor which is not the `html` element). You can remove one of them.

OTOH, I think it would be good to add an example to illustrate the "**child** text node" part of the rule.
```html
<body>
<div>
<p lang="en">Hello</p>
</div>
</body>
```
The `p` is good. The `div` is inapplicable. So this passes the rule.

Reply:
> Change the rule to "its closest ancestor with a lang attribute" or something similar

I think I'm actually tending towards this one, as the closest ancestor is indeed what plays the major part in the essence of this rule, as it relates more closely in how the language of the specific element itself is calculated/programmatically determined.

And mainly, because I do think there's value in making sure that it's clear that there can be failures of this rule that fail 3.1.2 but not 3.1.1.

Reply:
> These two examples are actually pretty much the same

ah, good spot. wanted to make the difference between parent and ancestor clear, but probably no need to do two examples that are so close to each other. will go with ancestor one, as parent is just a special case of ancestor

Reply:
Note that you can also go even simpler: 
check the existence of a `lang` attribute, not its validity. Then you can go for any ancestor without problem, the rule is much simpler and keep a single concern (it does not need to know what are valid/invalid language tags). Given that we already have rules that check validity of language tags, this is not going to be a real problem.

The point here being that we don't need the rule to be completely covering a single SC. It is better (imho) to have simple rules checking a simple thing and use a combination of rules to cover a SC (rules are already complicated beast, keeping them simple helps writers, readers and testers; simpler rules are also better for separation of concern).

This is a personal opinion and mostly showing what can be done. Of course, go the direction you want 😄 given that you are the one doing the job in the first place…

Reply:
i'll mull it over. appreciate the thorough discussion on this, particularly as i dove in on the deep end here :)